### PR TITLE
[4.0] Cleanup of override plugin

### PR DESCRIPTION
**Pull request by wilsonge**
- Small rework of the overrides installer plugin to use the built in class variables for app and db which makes typehints better and also removes our dependency on JFactory in the class. This also contains a bug fix (moving `$session->clear()` which clears the entire session contents - to `$session->remove()` which deletes the key we were passing in ~- clearing the entire session *might* be the cause of the session bug @brianteeman has been mentioning to me in the past)~

### CODE REVIEW

**Comment by brianteeman** [comment]
It really would be much better to have single issue commits. It makes testing MUCH easier and also prevents parts of this PR being held up by other parts

File: plugins/installer/override/override.php [hunk 1]
```diff
@@ -50,14 +50,19 @@ class PlgInstallerOverride extends CMSPlugin
 	 * @param   string  $name    The model name. Optional
 	 * @param   string  $prefix  The class prefix. Optional
 	 *
-	 * @return  \Joomla\CMS\MVC\Model\BaseDatabaseModel The model.
+	 * @return  \Joomla\Component\Templates\Administrator\Model\TemplateModel
 	 *
 	 * @since   4.0.0
+     *
+     * @throws \Exception
```
Comment (HLeithner) [inline plugins/installer/override/override.php]: CS

**Comment by mbabker** [comment]
> * moving `$session->clear()` which clears the entire session contents - to `$session->remove()` which deletes the key we were passing in

For the record while this is a correct change, there is a B/C layer in [`Joomla\CMS\Session\Session::clear()`](https://github.com/joomla/joomla-cms/blob/ad2c673ac27c2fd19bbdf6003aee0451d068de16/libraries/src/Session/Session.php#L204-L238) to cover this exact scenario so if the call to `$session->clear()` with arguments is nuking the entire session data store then this is a bug that should be fixed (preferably with automated tests but seems only a small few are able to provide tests for the 4.0 architecture).

**Comment by ghazal** [comment]
I have tested this item :red_circle: unsuccessfully on 76b89dca76869e2eda7e070c4dd26629947aa31e<br /><br />Got a fatal error:
Call to a member function getPaths() on null: Class 'Joomla\Plugin\System\Debug\Storage\FileStorage' not found
Latest Nightly Build (10/2/2019)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23840">issues.joomla.org/tracker/joomla-cms/23840</a>.</sub>

**Comment by wilsonge** [comment]
@ghazal you may need to delete the file `libraries/autoload_psr4.php` to force it to regenerate with the debug plugin

**Comment by brianteeman** [comment]
Can you look at resolving the conflicts so that this can be tested please. Or better still close it and submit new pr for each separate issue

**Comment by wilsonge** [comment]
Resolved. Moved debug stuff into a separate PR so this is a single issue pr now

File: plugins/installer/override/override.php [hunk 2]
```diff
@@ -311,14 +314,13 @@ public function load($id, $exid)
 	 *
 	 * @param   array  $pks  Updated files.
 	 *
-	 * @return  boolean
+	 * @return  void
 	 *
 	 * @since   4.0.0
+	 * throws   \Joomla\Database\Exception\ExecutionFailureException|\Joomla\Database\Exception\ConnectionFailureException
```
Comment (SharkyKZ) [inline plugins/installer/override/override.php]: Missing `@`.

**Comment by Schmidie64** [comment]
I have tested this item :white_check_mark: successfully on bd13b59ba4298fd63919bd190b48c7acbb099cb4<br /><br />as far as I can tell, everything continues to work as it should.

**System**
PHP Build On: Windows NT 10.0 build 18362 (Windows 10) i586
10.1.37-MariaDB
PHP-Version: 7.3.0
Webserver: Apache/2.4.37 (Win32) OpenSSL/1.1.1a PHP/7.3.0
Joomla! Version: 	Joomla! 4.0.0-alpha12-dev Development [ Amani ] 19-August-2019 13:40 GMT
User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.132 Safari/537.36<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23840">issues.joomla.org/tracker/joomla-cms/23840</a>.</sub>

**Comment by Schmidie64** [comment]
I have tested this item :white_check_mark: successfully on bd13b59ba4298fd63919bd190b48c7acbb099cb4<br /><br />I've tested this pr with the components from the Merged PR #21851 
as far as I can tell, everything continues to work as it should

**System**
PHP Build On: Windows NT 10.0 build 18362 (Windows 10) i586
10.1.37-MariaDB
PHP-Version: 7.3.0
Webserver: Apache/2.4.37 (Win32) OpenSSL/1.1.1a PHP/7.3.0
Joomla! Version: 	Joomla! 4.0.0-alpha12-dev Development [ Amani ] 19-August-2019 13:40 GMT
User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.132 Safari/537.36<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23840">issues.joomla.org/tracker/joomla-cms/23840</a>.</sub>

**Comment by bees4ever** [comment]
I have tested this item :white_check_mark: successfully on bd13b59ba4298fd63919bd190b48c7acbb099cb4<br /><br />For me it works as well.

## System
PHP 7.2.8 (built: Aug 28 2018 16:32:29) ( NTS )
Firefox 68 on macOS (Mojave)
Mac OS X 10.14.1 (18B75)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23840">issues.joomla.org/tracker/joomla-cms/23840</a>.</sub>

**Comment by ghost** [comment]
Status "Ready To Commit". 1 build failing.

**Comment by wilsonge** [comment]
Thanks

**Comment by Quy** [comment]
To reproduce:

* Get components from #21851
* Install com_helloworld.0.0.1
* Install com_helloworld.0.0.2

```
1 | () | JROOT\libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php:432
-- | -- | --
2 | Joomla\Database\Mysqli\MysqliStatement->execute() | JROOT\libraries\vendor\joomla\database\src\DatabaseDriver.php:673
3 | Joomla\Database\DatabaseDriver->execute() | JROOT\plugins\installer\override\override.php:407
4 | PlgInstallerOverride->saveOverrides() | JROOT\plugins\installer\override\override.php:191
5 | PlgInstallerOverride->finalize() | JROOT\plugins\installer\override\override.php:273
6 | PlgInstallerOverride->onInstallerAfterInstaller() | JROOT\libraries\src\Plugin\CMSPlugin.php:285
7 | Joomla\CMS\Plugin\CMSPlugin->Joomla\CMS\Plugin\{closure}() | JROOT\libraries\vendor\joomla\event\src\Dispatcher.php:495
8 | Joomla\Event\Dispatcher->dispatch() | JROOT\libraries\src\Application\EventAware.php:111
9 | Joomla\CMS\Application\WebApplication->triggerEvent() | JROOT\administrator\components\com_installer\Model\InstallModel.php:217
10 | Joomla\Component\Installer\Administrator\Model\InstallModel->install() | JROOT\administrator\components\com_installer\Controller\InstallController.php:42
11 | Joomla\Component\Installer\Administrator\Controller\InstallController->install() | JROOT\libraries\src\MVC\Controller\BaseController.php:728
12 | Joomla\CMS\MVC\Controller\BaseController->execute() | JROOT\libraries\src\Dispatcher\ComponentDispatcher.php:146
13 | Joomla\CMS\Dispatcher\ComponentDispatcher->dispatch() | JROOT\libraries\src\Component\ComponentHelper.php:383
14 | Joomla\CMS\Component\ComponentHelper::renderComponent() | JROOT\libraries\src\Application\AdministratorApplication.php:117
15 | Joomla\CMS\Application\AdministratorApplication->dispatch() | JROOT\libraries\src\Application\AdministratorApplication.php:160
16 | Joomla\CMS\Application\AdministratorApplication->doExecute() | JROOT\libraries\src\Application\CMSApplication.php:241
17 | Joomla\CMS\Application\CMSApplication->execute() | JROOT\administrator\includes\app.php:63
18 | require_once() | JROOT\administrator\index.php:36
```

PHP error log:
```
PHP Notice:  Undefined index: :preparedArray1 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 412
PHP Notice:  Undefined index: :preparedArray1 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 413
PHP Notice:  Undefined index: :preparedArray2 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 412
PHP Notice:  Undefined index: :preparedArray2 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 413
PHP Notice:  Undefined index: :preparedArray3 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 412
PHP Notice:  Undefined index: :preparedArray3 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 413
PHP Notice:  Undefined index: :preparedArray4 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 412
PHP Notice:  Undefined index: :preparedArray4 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 413
PHP Notice:  Undefined index: :preparedArray5 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 412
PHP Notice:  Undefined index: :preparedArray5 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 413
PHP Notice:  Undefined index: :preparedArray6 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 412
PHP Notice:  Undefined index: :preparedArray6 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 413
PHP Notice:  Undefined index: :preparedArray7 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 412
PHP Notice:  Undefined index: :preparedArray7 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 413
PHP Notice:  Undefined index: :preparedArray8 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 412
PHP Notice:  Undefined index: :preparedArray8 in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 413
PHP Warning:  mysqli_stmt::bind_param(): Number of variables doesn't match number of parameters in prepared statement in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 430
```

**Comment by SharkyKZ** [comment]
@Quy please test #26464.

**Pull request merged**